OpenRA.Mods.Common/Traits/Production/ProductionByDelivery.cs

```diff
@@ -0,0 +1,198 @@
+
+		public override object Create(ActorInitializer init) { return new ProductionByDelivery(init, this); }
+	}
+
+	class ProductionByDelivery : Production
+	{
+		readonly CPos startPos;
+		readonly CPos endPos;
+		readonly ProductionByDeliveryInfo info;
+
+		Actor self;
+		string factionVariant;
+
+		public ProductionByDelivery(ActorInitializer init, ProductionByDeliveryInfo info)
+			: base(init, info)
+		{
+			self = init.Self;
+
+			// Start a fixed distance away: the width of the map.
+			// This makes the production timing independent of spawnpoint
+			startPos = self.Location + new CVec(self.Owner.World.Map.Bounds.Width, 0);
+			endPos = new CPos(self.Owner.World.Map.Bounds.Left - 5, self.Location.Y);
+			this.info = info;
+		}
+
+		public override Actor DoProduction(Actor self, ActorInfo actorToProduce, ExitInfo exitInfo, string raceVariant)
+		{
+			var exit = self.Location + exitInfo.ExitCell;
+			var spawn = self.CenterPosition + exitInfo.SpawnOffset;
+
+			var td = new TypeDictionary
+			{
+				new OwnerInit(self.Owner),
+				new LocationInit(exit),
+				new CenterPositionInit(spawn)
+			};
+
+			if (!string.IsNullOrEmpty(raceVariant))
+				td.Add(new RaceInit(raceVariant));
+
+			return self.World.CreateActor(false, actorToProduce.Name, td);
+		}
+
+		public override bool Produce(Actor self, IEnumerable<ActorInfo> actorsToProduce, string raceVariant)
+		{
+			var deliveringActorType = info.DeliveryActor;
+			var owner = self.Owner;
+
+			this.self = self;
+			factionVariant = raceVariant;
+
+			// Check if there is a valid drop-off point before sending the transport
+			var exit = GetAvailableExit(self, actorsToProduce.First());
+			if (exit == null)
+				return false;
+
+			foreach (var trait in self.TraitsImplementing<INotifyDelivery>())
+				trait.IncomingDelivery(self);
+
+			owner.World.AddFrameEndTask(w =>
+			{
+				var altitude = self.World.Map.Rules.Actors[deliveringActorType].Traits.Get<PlaneInfo>().CruiseAltitude;
+				var deliveringActor = w.CreateActor(deliveringActorType, new TypeDictionary
+				{
+					new CenterPositionInit(w.Map.CenterOfCell(startPos) + new WVec(WRange.Zero, WRange.Zero, altitude)),
+					new OwnerInit(owner),
+					new FacingInit(64)
+				});
+
+				deliveringActor.QueueActivity(new Fly(deliveringActor, Target.FromCell(w, self.Location + new CVec(9, 0))));
+				deliveringActor.QueueActivity(new CallFunc(() => TryToLand(deliveringActor, actorsToProduce)));
+			});
+
+			return true;
+		}
+
+		void TryToLand(Actor deliveringActor, IEnumerable<ActorInfo> actorsToProduce)
+		{
+			// Check if there is a valid drop-off point before beginning to descend
+			var exit = GetAvailableExit(self, actorsToProduce.First());
+
+			// Abort the landing and refund the player
+			if (exit == null)
+			{
+				var value = actorsToProduce.Sum(actor => actor.Traits.Get<ValuedInfo>().Cost);
+				self.Owner.PlayerActor.Trait<PlayerResources>().GiveCash(value);
+
+				return;
```

Comment:
I may have missed it, but I think the `deliveringActor` doesn't get removed from the world in this case, does it?